consensus/fork_choice/tests/tests.rs

```diff
@@ -907,3 +922,254 @@ fn can_read_finalized_block() {
+            .unwrap()
+            .as_slice(),
+    );
+
+    let chain_config = ChainConfig {
+        weak_subjectivity_checkpoint: Some(Checkpoint { epoch, root }),
+        import_max_skip_slots: None,
+    };
+    let mut harness = ForkChoiceTest::new_with_chain_config(chain_config)
+        .apply_blocks(E::slots_per_epoch() as usize);
+
+    harness.harness.shutdown_receiver.close();
+    let msg = harness.harness.shutdown_receiver.try_next().unwrap();
+    assert!(msg.is_some());
+}
+
+#[test]
+fn weak_subjectivity_pass() {
+    let epoch = Epoch::new(0);
+    let root = Hash256::from_slice(
+        hex::decode("0000000000000000000000000000000000000000000000000000000000000000")
+            .unwrap()
+            .as_slice(),
+    );
+
+    let chain_config = ChainConfig {
+        weak_subjectivity_checkpoint: Some(Checkpoint { epoch, root }),
+        import_max_skip_slots: None,
+    };
+    let mut harness = ForkChoiceTest::new_with_chain_config(chain_config)
+        .apply_blocks(E::slots_per_epoch() as usize);
+
+    harness.harness.shutdown_receiver.close();
+    let msg = harness.harness.shutdown_receiver.try_next().unwrap();
+    assert!(msg.is_none());
+}
+
+#[test]
+fn weak_subjectivity_check_passes() {
+    let setup_harness = ForkChoiceTest::new()
+        .apply_blocks_while(|_, state| state.finalized_checkpoint.epoch == 0)
+        .apply_blocks(1)
+        .assert_finalized_epoch(2);
+
+    let checkpoint = setup_harness
+        .harness
+        .chain
+        .head_info()
+        .unwrap()
+        .finalized_checkpoint;
+
+    let chain_config = ChainConfig {
+        weak_subjectivity_checkpoint: Some(checkpoint),
+        import_max_skip_slots: None,
+    };
+
+    let mut test_harness = ForkChoiceTest::new_with_chain_config(chain_config.clone())
+        .apply_blocks_while(|_, state| state.finalized_checkpoint.epoch == 0)
+        .apply_blocks(1)
+        .assert_finalized_epoch(2);
+
+    test_harness.harness.shutdown_receiver.close();
+    let msg = test_harness.harness.shutdown_receiver.try_next().unwrap();
+    assert!(msg.is_none());
+}
+
+#[test]
+fn weak_subjectivity_check_fails_early_epoch() {
+    let setup_harness = ForkChoiceTest::new()
+        .apply_blocks_while(|_, state| state.finalized_checkpoint.epoch == 0)
+        .apply_blocks(1)
+        .assert_finalized_epoch(2);
+
+    let mut checkpoint = setup_harness
+        .harness
+        .chain
+        .head_info()
+        .unwrap()
+        .finalized_checkpoint;
+
+    checkpoint.epoch = checkpoint.epoch - 1;
+
+    let chain_config = ChainConfig {
+        weak_subjectivity_checkpoint: Some(checkpoint),
+        import_max_skip_slots: None,
+    };
+
+    let mut test_harness = ForkChoiceTest::new_with_chain_config(chain_config.clone())
+        .apply_blocks_while(|_, state| state.finalized_checkpoint.epoch == 0)
+        .apply_blocks(1)
+        .assert_finalized_epoch(2);
+
```

Comment:
Could we also throw this check in there?

```rust
assert!(harness.chain.head_info().unwrap().slot < checkpoint.epoch.start_slot(E::slots_per_epoch()));
```

Reply:
do you mean head_slot > checkpoint.epoch.start_slot?